Fixed Settings Screen Tests

CananArikan opened the pull request:
testHideShowToggle is fixed, closes #480 
The parent issue to #480 is #405 

# Description of Work
Fixed broken settings screen test. This is part of the effort to unlink the chained tests.

CananArikan committed:
Fixed Settings Screen Tests

testHideShowToggle is fixed.

CananArikan committed:
Updated PR

Fixed the test for iPhone.

CananArikan committed:
Merge branch 'develop' into test/480-fix-setting-screen-tests

CananArikan committed:
Disabled tests

We are disabling testReorder and testAddCustomCategory, they will be updated after the Update Category List UI is completed: #492

rhsalas3 committed:
Fixed failing CustomCategoriesTests; disabled expected tests

Clstroud commented on VocableUITests/Tests/CustomCategoriesTests.swift:
```diff
@@ -13,6 +13,7 @@ class CustomCategoriesTests: CustomCategoriesBaseTest {
     func testAddNewPhrase() {
         let customPhrase = "dd"
         let confirmationAlert = "Are you sure? Going back before saving will clear any edits made."
```
We should probably give this an identifier instead of keying off of the actual string, since this is subject to change and would break the test?

Clstroud replied:
Looks like that identifier is currently `text_editor.alert.cancel_editing_confirmation.title`

Clstroud replied:
Just realized this ^ may not actually be the accessibility ID, so I might've accidentally lied there. It's definitely the localization key, and I'm thinking we may want to make them match up for elements like this.

CananArikan replied:
Fixed!

Clstroud commented on VocableUITests/Tests/SettingsScreenTests.swift:
```diff
@@ -3,45 +3,42 @@
+        settingsScreen.showCategorySwitch.tap()
+        settingsScreen.leaveCategoryDetailButton.tap()
+        
+        settingsScreen.locateCategoryCell(category)
+        XCTAssertTrue(settingsScreen.locateCategoryCell(category).buttons[settingsScreen.categoryUpButton].isEnabled)
+        XCTAssertTrue(settingsScreen.locateCategoryCell(category).buttons[settingsScreen.categoryDownButton].isEnabled)
+        XCTAssertTrue(settingsScreen.locateCategoryCell(category).buttons[settingsScreen.categoryForwardButton].isEnabled)
```
```suggestion
        let categoryCell = settingsScreen.locateCategoryCell(category)
        guard categoryCell.waitForExistence(20) else 
            XCTFail("Couldn't locate category cell")
            return
        }        
        XCTAssertTrue(categoryCell.buttons[settingsScreen.categoryUpButton].isEnabled)
        XCTAssertTrue(categoryCell.buttons[settingsScreen.categoryDownButton].isEnabled)
        XCTAssertTrue(categoryCell.buttons[settingsScreen.categoryForwardButton].isEnabled)
```
Just using this one snippet as an example, I think you can grab the cell element once and store it as a variable to reference in the assertions below it. I also threw in a `waitForExistence` because I'm not sure exactly what happens if `locateCategoryCell()` is unsuccessful, but [there are other strategies](https://mokacoding.com/blog/write-better-swift-tests-with-xctest-assertions/) that might help with that. Not a huge deal, mostly just wanted to grab the cell once for performance and readability reasons 😄

CananArikan replied:
We try to “handle" what happens if the cell is not present initially in the loop within `locateCategoryCell()`. We don't throw a failure explicitly but we do let the query fail which will report that the cell could not be found. Before that, however, we scroll the pages within the list of categories to find the cell we're looking for. 

Making a reference to the cell once in order to shorten our run time and the line of code is a good idea. We will implement that.

CananArikan replied:
Fixed.

Clstroud commented on VocableUITests/Screens/SettingsScreen.swift:
```diff
@@ -17,39 +18,51 @@ class SettingsScreen: BaseScreen {
+        return categoryCellQuery(category)
+    }
+    
+    func categoryCellQuery(_ category: String) -> XCUIElementQuery {
```
Does this get used outside this context? I wonder if maybe it can be `private`?

CananArikan replied:
We were thinking that we would use it outside of this context when we fix some of the other tests. But right now it is just being used here. I think we will go ahead and make it private. We can always change it back when we need to 😉

CananArikan replied:
We updated it!

Clstroud commented on VocableUITests/Tests/SettingsScreenTests.swift:
```diff
@@ -54,18 +51,19 @@ class SettingsScreenTests: BaseTest {
         XCTAssert(settingsScreen.otherElements.containing(.staticText, identifier: generalCategoryText).element.exists)
         XCTAssert(settingsScreen.otherElements.containing(.staticText, identifier: basicNeedsCategoryText).element.exists)
```
Based on some of the test failures I've seen recently, we should probably be using `waitForExistence()` to handle slow transition animations. I'm worried this could potentially be flaky otherwise, but I might be missing some context.

Clstroud replied:
Also: same note about storing elements as variables. `settingsScreen.otherElements.containing(.staticText, identifier: generalCategoryText).element` is _very_ long and difficult to read multiple times 😅

CananArikan replied:
We disabled this test for now, the code you pointed out is something we plan to refactor but we are waiting for #492 to be completed because of the Categories UI changes. When that ticket is completed we plan to update the flow in this test.

Clstroud reviewed (changes requested):
Structurally this seems pretty solid, but I left some feedback notes for tightening things up a bit. Happy to discuss any/all suggestions 😄

CananArikan committed:
Fixed PR

As recommended, the cell element is stored as a variable to reference in the assertions.

CananArikan committed:
Added alert identifier

Assigned an identifier to the alert message description and refactored test queries to use the identifier.

Clstroud reviewed (approved):
🏆 💯 🥇 Nice updates!
